forcetree.c

```diff
@@ -144,57 +162,134 @@ int force_tree_build(int npart)
-            subnode += 2;
-        if(P[p_i].Pos[2] > node->center[2])
-            subnode += 4;
-    }
+    int subnode = 0;
+    int k;
+    for(k=0; k<3; k++)
+        subnode |= (P[p_i].Pos[k] > node->center[k]) << k;
+    return subnode;
+}
 
-    if(!All.NoTreeRnd && node->len < 1.0e-3 * All.ForceSoftening[P[p_i].Type])
-    {
-        /* seems like we're dealing with particles at identical (or extremely close)
-         * locations. Randomize subnode index to allow tree construction. Note: Multipole moments
-         * of tree are still correct, but this will only happen well below gravitational softening
-         * length-scale anyway.
-         */
-        return (int) (7.99 * get_random_number(P[p_i].ID+nodepos));
+/*Check whether a particle is inside the volume covered by a node*/
+inline int inside_node(const struct NODE * node, const int p_i)
+{
+    int k;
+    for(k=0; k<3; k++) {
+        double pdiff = 2*(P[p_i].Pos[k] - node->center[k]);
+        pdiff = pdiff > 0 ? pdiff : -1*pdiff;
+        if(pdiff > node->len)
+            return 0;
     }
-    return subnode;
+    return 1;
+}
+
+/*Initialise an internal node at nfreep. The parent is assumed to be locked, and
+ * we have assured that nothing else will change nfreep while we are here.*/
+static void init_internal_node(struct NODE *nfreep, struct NODE *parent, int subnode)
+{
+            int j;
+            const MyFloat lenhalf = 0.25 * parent->len;
+
+            nfreep->len = 0.5 * parent->len;
+
+            for(j = 0; j < 3; j++) {
+                /* Detect which quadrant we are in by testing the bits of subnode:
+                 * if (subnode & [1,2,4]) is true we add lenhalf, otherwise subtract lenhalf*/
+                const int sign = (subnode & (1 << j)) ? 1 : -1;
+                nfreep->center[j] = parent->center[j] + sign*lenhalf;
+            }
+            for(j = 0; j < 8; j++)
+                nfreep->u.suns[j] = -1;
 }
 
 /* Size of the free Node thread cache.
  * 100 was found to be optimal for an Intel skylake with 4 threads.*/
 #define NODECACHE_SIZE 100
 
-/*! Constructs the gravitational oct-tree.
- *
- *  The index convention for accessing tree nodes is the following: the
- *  indices 0...NumPart-1 reference single particles, the indices
- *  All.MaxPart.... All.MaxPart+nodes-1 reference tree nodes. `Nodes_base'
- *  points to the first tree node, while `nodes' is shifted such that
- *  nodes[All.MaxPart] gives the first tree node. Finally, node indices
- *  with values 'All.MaxPart + MaxNodes' and larger indicate "pseudo
- *  particles", i.e. multipole moments of top-level nodes that lie on
- *  different CPUs. If such a node needs to be opened, the corresponding
- *  particle must be exported to that CPU. */
-int force_tree_build_single(int npart)
+/*Get a pointer to memory for a free node, from our node cache.
+ * If there is no memory left, return NULL.*/
+int get_freenode(int * nfree, int *nfree_thread, int *numfree_thread)
```

Comment:
nfree_thread and numfree_thread looked confusing... Is there a better name?

Reply:
nnext and nrem?